src/vmm/src/default_syscalls/filters.rs

```diff
@@ -1,20 +1,83 @@
 // Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
-use std::convert::TryInto;
-
 use seccomp::{
-    allow_syscall, allow_syscall_if, BpfProgram, Error, SeccompAction, SeccompCmpArgLen as ArgLen,
-    SeccompCmpOp::Eq, SeccompCondition as Cond, SeccompError, SeccompFilter, SeccompLevel,
-    SeccompRule,
+    allow_syscall, allow_syscall_if, deserialize_binary, BpfProgram, BpfThreadMap,
+    DeserializationError, Error, SeccompAction, SeccompCmpArgLen as ArgLen, SeccompCmpOp::Eq,
+    SeccompCondition as Cond, SeccompFilter, SeccompRule,
 };
+use std::fmt;
+use std::fs::File;
 use utils::signal::sigrtmin;
 
-/// The default filter containing the white listed syscall rules required by `Firecracker` to
+const THREAD_CATEGORIES: [&str; 3] = ["vmm", "api", "vcpu"];
+
+/// Error retrieving seccomp filters.
+#[derive(fmt::Debug)]
+pub enum FilterError {
+    /// Filter deserialitaion error.
+    Deserialization(DeserializationError),
+    /// Invalid thread categories.
+    ThreadCategories(String),
+    /// Missing Thread Category.
+    MissingThreadCategory(String),
+    /// Seccomp error occurred.
+    Seccomp(Error),
+}
+
+impl fmt::Display for FilterError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::FilterError::*;
+
+        match *self {
+            Deserialization(ref err) => write!(f, "Filter (de)serialization failed: {}", err),
+            ThreadCategories(ref categories) => {
+                write!(f, "Invalid thread categories: {}", categories)
+            }
+            MissingThreadCategory(ref category) => {
+                write!(f, "Missing thread category: {}", category)
+            }
+            Seccomp(ref err) => write!(f, "Seccomp error: {}", err),
+        }
+    }
+}
+
+/// Return an error if the BpfThreadMap contains invalid thread categories.
+fn filter_thread_categories(map: BpfThreadMap) -> Result<BpfThreadMap, FilterError> {
```

Comment:
nit:
This seems like more handling than is needed. If the `BpfThreadMap` has entries responsible for other thread categories, or lacks mandatory thread categories, we can simply return that the expected thread categories are `vmm`,`api` and `vcpu`.

I feel that the level of handling here implies logic which sets the expectation for more and more error reporting, which is not needed IMO.